PWGHF/TableProducer/candidateCreatorB0.cxx

```diff
@@ -376,41 +355,95 @@ struct HfCandidateCreatorB0Expressions {
-      auto arrayDaughters = array{candD.prong0_as<aod::BigTracksMC>(),
-                                  candD.prong1_as<aod::BigTracksMC>(),
-                                  candD.prong2_as<aod::BigTracksMC>(),
-                                  candidate.prong1_as<aod::BigTracksMC>()};
+      auto arrayDaughtersB0 = array{candD.prong0_as<aod::BigTracksMC>(),
+                                    candD.prong1_as<aod::BigTracksMC>(),
+                                    candD.prong2_as<aod::BigTracksMC>(),
+                                    candidate.prong1_as<aod::BigTracksMC>()};
       auto arrayDaughtersD = array{candD.prong0_as<aod::BigTracksMC>(),
                                    candD.prong1_as<aod::BigTracksMC>(),
                                    candD.prong2_as<aod::BigTracksMC>()};
+
       // B0 → D- π+ → (π- K+ π-) π+
       // Printf("Checking B0 → D- π+");
-      indexRec = RecoDecay::getMatchedMCRec(particlesMC, arrayDaughters, pdg::Code::kB0, array{-kPiPlus, +kKPlus, -kPiPlus, +kPiPlus}, true, &sign, 2);
+      indexRec = RecoDecay::getMatchedMCRec(particlesMc, arrayDaughtersB0, pdg::Code::kB0, array{-kPiPlus, +kKPlus, -kPiPlus, +kPiPlus}, true, &sign, 3);
       if (indexRec > -1) {
         // D- → π- K+ π-
         // Printf("Checking D- → π- K+ π-");
-        indexRec = RecoDecay::getMatchedMCRec(particlesMC, arrayDaughtersD, pdg::Code::kDMinus, array{-kPiPlus, +kKPlus, -kPiPlus}, true, &sign, 2);
+        indexRec = RecoDecay::getMatchedMCRec(particlesMc, arrayDaughtersD, pdg::Code::kDMinus, array{-kPiPlus, +kKPlus, -kPiPlus}, true, &sign, 2);
         if (indexRec > -1) {
           flag = sign * BIT(hf_cand_b0::DecayType::B0ToDPi);
         } else {
           debug = 1;
           LOGF(info, "WARNING: B0 in decays in the expected final state but the condition on the intermediate state is not fulfilled");
         }
       }
+
+      // B0 → Ds- π+ → (K- K+ π-) π+
+      if (!flag) {
+        indexRec = RecoDecay::getMatchedMCRec(particlesMc, arrayDaughtersB0, pdg::Code::kB0, array{-kKPlus, +kKPlus, -kPiPlus, +kPiPlus}, true, &sign, 3);
+        if (indexRec > -1) {
+          // Ds- → K- K+ π-
+          indexRec = RecoDecay::getMatchedMCRec(particlesMc, arrayDaughtersD, -pdg::Code::kDS, array{-kKPlus, +kKPlus, -kPiPlus}, true, &sign, 2);
+          if (indexRec > -1) {
+            flag = sign * BIT(hf_cand_b0::DecayTypeMc::B0ToDsPiAndDsToKKPi);
+          }
+        }
+      }
+
+      // Partly reconstructed decays, i.e. the 4 prongs have a common b-hadron ancestor
+      // convention: final state particles are prong0,1,2,3
+      if (!flag) {
+        auto particleProng0 = arrayDaughtersB0[0].mcParticle_as<aod::McParticles>();
+        auto particleProng1 = arrayDaughtersB0[1].mcParticle_as<aod::McParticles>();
+        auto particleProng2 = arrayDaughtersB0[2].mcParticle_as<aod::McParticles>();
+        auto particleProng3 = arrayDaughtersB0[3].mcParticle_as<aod::McParticles>();
```

Comment:
not sure but maybe this works also without cast (i.e. simply `.mcParticle()` instead of `mcParticle_as<aod::McParticles>()`?

Reply:
Indeed, it should be equivalent according to [this line](https://github.com/AliceO2Group/AliceO2/blob/dev/Framework/Core/include/Framework/AnalysisDataModel.h#L1412), and I works with this modification on my side.